### PR TITLE
MultiSig Implementation

### LINKED ISSUE #360: Multisig Account
Create MultiSig functionality to allow multiple signatories for creating a multisig address/bytes.

### PR DESCRIPTION
**Pull request by justkawal**
## **User description**
MultiSig
  - Implements MultiSig Address Creation
      - ExcludeMe method from sorted address
  - Implement Calls for multisig
      -  `ApproveAsMulti`
      - `AsMulti`
      - `CancelAsMulti`
  - Test Cases
  - Closes #360

___

## **Type**
Enhancement, Tests

___

## **Description**
- Implemented MultiSig address creation functionality in the `polkadart_keyring` package.
- Added a new class `MultiSig` with methods for creating multisig bytes and multisig addresses.
- Added several utility functions: `bnToU8a`, `_bigIntToBytes`, and `uint8ListCompare`.
- Added a new enum `KeyPairType` with three types: `ed25519`, `sr25519`, and `ecdsa`.
- Added a new function `blake2bDigest` for hashing data.
- Added two new test cases for validating the creation of multikey and multikey address.
- Updated README with MultiSig address creation example.
- Added a new dependency `polkadart_scale_codec: ^1.1.2`.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>polkadart_keyring.dart</strong><dd><code>Reorganization of library structure and addition of new imports</code>&nbsp; &nbsp; </dd></summary>
<hr>

packages/polkadart_keyring/lib/polkadart_keyring.dart
- Added import for 'dart:convert' and <br>  'package:polkadart_scale_codec/polkadart_scale_codec.dart'.
- Reorganized the structure of the library by separating source code and <br>  utility files.
- Added 'src/multisig.dart' to the part directive.
- Moved 'src/extensions.dart' and 'src/constants.dart' to the 'utils' <br>  directory.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/417/files#diff-d094bf0f2dcb6f81b7c2a1501e653f30fc331514035b7ef2455334cf680c0962">+11/-2</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>analysis_options.yaml</strong><dd><code>Addition of new linting rule</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/polkadart_keyring/analysis_options.yaml
- Added a new rule `non_constant_identifier_names: false`.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/417/files#diff-7883da93f53414b52553b9012be17e01775baaf0b1f960b7dea8068fd4a1a49c">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>multisig.dart</strong><dd><code>Implementation of MultiSig class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/polkadart_keyring/lib/src/multisig.dart
- Added a new class `MultiSig` with methods for creating multisig bytes <br>  and multisig addresses.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/417/files#diff-6f22452466e8a7679c95f54554c65115f50582d68868dbe8004f280ecbd068be">+43/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>constants.dart</strong><dd><code>Addition of new constant</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/polkadart_keyring/lib/utils/constants.dart
- Added a constant `DEV_PHRASE`.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/417/files#diff-7f159ca737e2da2d423fde84f3b16b0f6f9e6bd2ed495378077abe9c19b39cee">+5/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>extensions.dart</strong><dd><code>Addition of KeyPairType enum</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/polkadart_keyring/lib/utils/extensions.dart
- Added a new enum `KeyPairType` with three types: `ed25519`, `sr25519`, `<br>  `and `<br>  ``ecdsa`.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/417/files#diff-e294363fba8b2b6abbf4410dcc1e3aff5585fa7a3636b6368801a316d10af75b">+14/-1</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>hashers.dart</strong><dd><code>Addition of new hashing function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/polkadart_keyring/lib/utils/hashers.dart
- Added a new function `blake2bDigest` for hashing data.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/417/files#diff-3a4264fc9d90d76e0afe036530678889ac95c997798aa32c8ffc6259f519a47b">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>utilities.dart</strong><dd><code>Addition of new utility functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/polkadart_keyring/lib/utils/utilities.dart
- Added several utility functions: `bnToU8a`, `_bigIntToBytes`, and `<br>  ``uint8ListCompare`.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/417/files#diff-e018461e3ffdc7dbdfb8fb7133d0dcdc4a909836363b3eb7d4611cf3a3b7451a">+59/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>multisig_test.dart</strong><dd><code>Addition of new test cases for MultiSig</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/polkadart_keyring/test/multisig_test.dart
- Added two new test cases for validating the creation of multikey and <br>  multikey address.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/417/files#diff-87c6cd89b664db4b287c43cb0438e247a72c908f2c5489c038357a16123559b8">+60/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Updated README with MultiSig address creation example</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/polkadart_keyring/README.md
- Added a new section for creating MultiSig address.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/417/files#diff-5d8253fa6ea0ba59fa414776d193f7c630f4f35b21c794c6263bb4e57918ba33">+11/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Dependencies</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>pubspec.yaml</strong><dd><code>Addition of new dependency</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/polkadart_keyring/pubspec.yaml
- Added a new dependency `polkadart_scale_codec: ^1.1.2`.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/417/files#diff-fa065893043dfdd83b00f41c638e8bbe7a9a4d8a202bcb7f3993924e5431b94d">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Inline File Walkthrough 💎</strong></summary><hr>

For enhanced user experience, the `describe` tool can add file summaries directly to the "Files changed" tab in the PR page.
This will enable you to quickly understand the changes in each file, while reviewing the code changes (diffs).

To enable inline file summary, set `pr_description.inline_file_summary` in the configuration file, possible values are:
- `'table'`: File changes walkthrough table will be displayed on the top of the "Files changed" tab, in addition to the "Conversation" tab.
- `true`: A collapsable file comment with changes title and a changes summary for each file in the PR.
- `false` (default): File changes walkthrough will be added only to the "Conversation" tab.
<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

File: packages/polkadart_keyring/README.md [hunk 1]
```diff
@@ -13,6 +13,17 @@ dependencies:
...
 ## Usage
 
+### Creating MultiSig Address
+
+```dart
+    final result = MultiSig.createMultiSigAddress([
+        '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY',
+        '5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty',
+        '5FLSigC9HGRKVhB9FiEo4Y3koPsNmBmLJbpXg2mp1hXcS59Y'
+        ], 2);
```
Comment (leonardocustodio) [inline packages/polkadart_keyring/README.md:23]: Would be nice to put here in the docs that 2 is the threshold and what it means, here you can find a reference text: https://wiki.polkadot.network/docs/learn-account-multisig

**Review by leonardocustodio — CHANGES_REQUESTED** [review]
Also, it is important to note that the scope of this ticket isn't only multisig account creation. We also need to be able to send the call, decode the call, make the deposit, cancel the call.

You can find more information in the following links:
https://wiki.polkadot.network/docs/learn-guides-accounts-multisig
https://support.polkadot.network/support/solutions/articles/65000181826-how-to-create-and-use-a-multisig-account

**Comment by justkawal** [comment]
Looking into it.

**Comment by justkawal** [comment]
> Also, it is important to note that the scope of this ticket isn't only multisig account creation. We also need to be able to send the call, decode the call, make the deposit, cancel the call.
> 
> You can find more information in the following links: https://wiki.polkadot.network/docs/learn-guides-accounts-multisig https://support.polkadot.network/support/solutions/articles/65000181826-how-to-create-and-use-a-multisig-account

I researched the internet, but wasn't able to find a code sample or example to mimic/implement the multisig (`initiate`, `approveAsMulti`, `asMulti`, and `cancel`).

Whenever you find time, let me know. @leonardocustodio

**Comment by leonardocustodio** [comment]
You can find the methods here: https://github.com/paritytech/txwrapper-core/tree/main/packages/txwrapper-substrate/src/methods/multisig

**Comment by justkawal** [comment]
@leonardocustodio I guess the PR looks good now....

**Comment by leonardocustodio** [comment]
Is this also for #359 or only multisig?

**Comment by justkawal** [comment]
> Is this also for #359 or only multisig?

Umm.... Yes,
I have added functionality to get and parse every `signedExtensions` from the `chain metadata` even the `custom` as well and provide access to that from `TypeRegistry`....

**Review by leonardocustodio — APPROVED** [review]
lgtm

**Pull request merged**